File: 3782.last-remaining-integer-after-alternating-deletion-operations.go
#
# @lc app=leetcode id=3782 lang=golang
#
# [3782] Last Remaining Integer After Alternating Deletion Operations
#

# @lc code=start
func lastInteger(n int64) int64 {
	lengthHistory := []int64{n}
	cur := n
	for cur > 1 {
		cur = (cur + 1) / 2
		lengthHistory = append(lengthHistory, cur)
	}
	pos := int64(1)
	for i := len(lengthHistory) - 1; i >= 1; i-- {
		prevLen := lengthHistory[i-1]
		step := i
		if step%2 == 1 {
			pos = 2*pos - 1
		} else {
			if prevLen%2 == 0 {
				pos = 2 * pos
			} else {
				pos = 2*pos - 1
			}
		}
	}
	return pos
}
# @lc code=end

File: 3782.last-remaining-integer-after-alternating-deletion-operations.java
#
# @lc app=leetcode id=3782 lang=java
#
# [3782] Last Remaining Integer After Alternating Deletion Operations
#

# @lc code=start
class Solution {
  public long lastInteger(long n) {
    long start = 1L;
    long step = 1L;
    long length = n;
    boolean isLeftTurn = true;
    while (length > 1) {
      if (isLeftTurn) {
        step *= 2L;
      } else {
        if (length % 2L == 1L) {
          step *= 2L;
        } else {
          start += step;
          step *= 2L;
        }
      }
      length = (length + 1L) / 2L;
      isLeftTurn = !isLeftTurn;
    }
    return start;
  }
}
# @lc code=end